### DISCUSSION TITLE
Training Custom Models

### DISCUSSION THREAD

**Discussion by Mattk70** [Mattk70/Chirpity-Electron#330] [Show and tell]
With Chirpity 5, members can train and load custom models (classifiers). This post gives an end-to-end example of how to train, load and use a custom model.

**_Updated for 5.1.0_**

For this example, we are going to train a model to recognise a variety of European amphibians (frogs and toads specifically) and combine this with BirdNET, so BirdNET predictions are aware of these amphibians.

1. Firstly, I have assembled a dataset of Frogs audio from Xeno-Canto, essentially [this](https://xeno-canto.org/explore?query=cnt:%22%3DFrance%22%20grp:%22frogs%22%20type:%27call%27) The files for each species are collected into a separate folder, so they can be analysed species by species.

2. To generate the audio used for training we need to create 3 second audio segments which contain the species' calls.  To generate these segments, I ran a BirdNET analysis of these files, species by species, using the "Everything" list. 

3. Since BirdNET does not "know" about these species, it gives a variety of incorrect IDs. Some are the frog calls, others are not: I filtered the analysis results by each species detected to establish if what BirdNET thought was a bird/frog/engine was in fact a Frog or Toad sound.

4. If the IDs appeared to be amphibian sounds, I'd export the audio from the summary of that species ID to a folder for the frog species I'd analysed, matching BirdNET's original label format:

![image](https://github.com/user-attachments/assets/26e61cad-ec31-4402-be0b-0a70aef6825b)

The eagle-eyed will have noticed there is a folder named "Background_Background". This contains a good selection of audio that is not a Frog/Toad call. 

>[!tip] 
>Whenever you train a model, having a good selection of background sounds is essential for decent performance. If you don't, every sound in the world presented to the model will be classified as one of the target species - in this case a Frog or Toad.

There are also a few bird species. These are included as species which regularly appear as false positives in BirdNET analyses where amphibians are calling. I include these so the new model will learn to distinguish between them better.

>[!important]
> These bird species have a '-' at the end of their common name so that name is different from the existing BirdNET label. All labels MUST be unique.

5. Now the fun part, let's train a model! 
>[!note]
> The option to train a custom model is only active when you have selected the BirdNET model in Settings

![image](https://github.com/user-attachments/assets/f58977bb-3f0f-4cae-8671-d3668784b30c)

Selecting this menu item will open the model training form:

![image](https://github.com/user-attachments/assets/a60a738c-317e-4df2-8b8c-3adaa28ad61f)

6. **Training Audio Location:** Select the location you saved your training audio files.
7.  **Dataset Cache Location:** In order that training runs efficiently, the audio is prepared into one big binary with labels, so you can choose the location to save a cache of that prepared audio. By default, it will be saved to the **Training Audio Location**
>[!tip]
> Choose a location with ~~lots of free space, and~~ fast I/O (avoid network locations and USB devices if you can)

>[!note]
> "Replace labels" will give you a model that only knows about your saved species, "Append labels" will add your labels to those of BirdNET. 

9. Now you've selected your training folder and cache location, you can use the default values so just hit "Start training"

>[!important]
> Training will split your data into a set of examples to train on and another set to hold back for validation. The validation examples are not seen during training - they are used to measure how performance for unseen examples improves during the training run. Because of this, you need at least two examples for each species / class in your data.

10. Grab a coffee while the Chirpity prepares your data, and runs the training. It will take some time. For context my high end machine took ~~c. 45 minutes~~ just over 4 minutes to train this model with c. 6100 examples in the dataset. 

>[!tip]
> Start small - because you are training on top of what BirdNET has already learned, only a few good examples are needed (about 100 or so). Try to include examples that reflect a variety of calls and vocalisations across recording conditions (e.g. weather, habitats, close/distant). Fundamentally, the quality of the data makes the difference between an average model and an excellent one.

>[!note]
>During training, the footer will show a progress bar, and at the end of each Epoch, the latest loss and accuracy figures are reported:

![image](https://github.com/user-attachments/assets/358e5015-a32c-4925-bd36-2a267d968436)

You are looking for validation accuracy to increase over time, and validation loss to decrease. As noted before, the validation figures reflect model performance on unseen examples.

When the model has trained for the number of  epochs you set, or if performance is not improving, training will end and the new model will be saved in your chosen location. You'll then see an interactive chart of the training history:

![image](https://github.com/user-attachments/assets/ae8d8005-583d-493a-872c-5bdea2de23a6)

>[!tip] 
> As you can see in this chart, most of the learning happened in the first 10 Epochs, with only very minor improvements after that. However the shape of the gradients show stable performance and a good result. Marked fluctuations would indicate instability. Worst of all, poor data or settings choices can lead to NaN values. If you see NaN for loss, just abort the training (hit Escape) and try different settings or refine your training data.

### Loading your model

Once the model has been saved, you can go back to the Train menu and select "Import a Custom Model". Give it a name, and select the folder where it has been saved. Your new model will be added to the "Models" list in settings 

>[!tip]
>New classes will appear in the "Everything" lists. They will not appear in the other lists (if you added 50 classes or less, they will - by chance - also appear in the "All Birds" list. More than that and some will be missing). If you want to filter the species in a model which has classes "appended" to BirdNET, use a custom list.

**Comment by Lei-Jue** [comment]
Hello Matthew,

Thank you for developing the custom model training feature in Chirpity.

I recently trained a model to detect the Maritime Striped Squirrel (Tamiops maritimus) in forest environments. The training process went smoothly, and the model achieved an accuracy of 88.61%.

However, when I deployed the custom model (with everything list), I noticed that it detected squirrels far more frequently than other bird species—sometimes dozens of times more. While it does identify some true positives, there are also many false positives with high confidence scores.

Here’s some background on my dataset: I used 11 audio recordings of Tamiops maritimus from the Xeno-canto website and field recordings. The audio files were split into 3-second snippets, then organized into two folders with manually reviewed: "Tamiops maritimus_Maritime Striped Squirrel" and "Background_Background".

I’d like to understand why the model is producing so many irrelevant results classified as squirrels, and how I can improve its performance.

Thanks again!

<img width="554" height="287" alt="2" src="https://github.com/user-attachments/assets/b7c039c7-33a3-491f-9410-8d2623351ec9" />
<img width="554" height="386" alt="1" src="https://github.com/user-attachments/assets/fb14da13-66c3-459a-8c83-89e2695da66c" />

**Comment by Mattk70** [comment]
Hi @Lei-Jue ,

There are many factors that influence classifier performance. From what you have shown and said, I would make these observations:

1. The 88.6% figure is how well the model classifies the training data. The key metric is to look at validation data, which is how well it classifies unseen examples - and therefore generalises -  and this is 83%.  83% is not great. The model makes 1200 predictions per hour, so at 83% accuracy, you can expect at least 204 incorrect predictions per hour of analysis. It will be more, since in the real world there will be other sounds you haven't trained on, and it will classify these as one of your training classes. A well-performing classifier will have > 95% validation accuracy.
2. 11 training examples is possibly too few to allow the classifier to learn the essence of the squirrel call structure. Use a variety of calls, ideally in different environments, and include both clear and distant calls.
3. Use _a lot_ of background examples. Make sure these are both similar to the background of your training data, and the environment where you are recording.
4. Model training is iterative. It takes many runs to get the best result. Start with simple training settings and test one parameter change at a time. Optimising learning rate, loss function and augmentations used can make a big difference.

It will help to ask an AI chatbot about tips for your specific situation, I have found them to give good advice.
